src/BinWrapper.jl

```diff
@@ -0,0 +1,161 @@
+    close(cmdin)
+    cmdin = open(cmdin_p)
+
+    # run a process
+    process = run(pipeline(ignorestatus(cmd), stdin=cmdin, stdout=cmdout, stderr=cmderr))
+
+    # read out/err of a command
+    cmdout = open(cmdout_p)
+    cmderr = open(cmderr_p)
+
+    out = read(cmdout, String)
+    err = read(cmderr, String)
+
+    # return all the needed data
+    return (
+      out     = out,
+      err     = err,
+      ret     = process.exitcode
+    )
+  else # non-tempfile-based implementation
+    # create pipes/buffers for in/out/err of a command
+    cmdin  = IOBuffer(input)
+    cmdout = IOBuffer()
+    cmderr = IOBuffer()
+
+    if VERSION < v"1.1"
+      cmdin  = Pipe()
+      cmdout = Pipe()
+      cmderr = Pipe()
+    end
+
+    # run a process
+    process = run(pipeline(ignorestatus(cmd), stdin=cmdin, stdout=cmdout, stderr=cmderr), wait=!(VERSION < v"1.1"))
+
+    if VERSION < v"1.1"
+      # close unnecessary communication tunnels of pipes initialized by pipeline()
+      close(cmdout.in)
+      close(cmderr.in)
+
+      # pass the input to the process and close the related pipe
+      # note: this could be also `write(process, input)`, but then `cmdin` would not be used which is not cool
+      write(cmdin, input)
+      close(cmdin)
+
+      # asynchronously read the outputs of the command
+      out = @async String(read(cmdout))
+      err = @async String(read(cmderr))
+
+      # return all the needed data
+      return (
+        out     = fetch(out),
+        err     = fetch(err),
+        ret     = process.exitcode
+      )
+    end # if VERSION < v"1.1"
+
+    # read the outputs of the command
+    out = String(take!(cmdout))
+    err = String(take!(cmderr))
+
+    # return all the needed data
+    return (
+      out     = out,
+      err     = err,
+      ret     = process.exitcode
+    )
+  end
+end
+
+# This function wraps the binaries that come with the original kbmag library.
+# Here, all the parameters are passed as Strings
+function kbmag_bin_wrapper(program::String; params = (), input::String = "", force_dir::String = ".")
+  # remember the current working directory
+  call_dir = pwd()
+
+  # change the directory if necessary
+  # WARNING: when calling a kbmag binary we need to make sure that
+  # all the input files are in the current working directory
+  if force_dir != "."
+    cd(force_dir)
+  end
+
+  result = low_level_exec(`$kbmag_bin_dir$program $(params)`; input = input)
+
+  # change the current working directory back to the previous one
+  if force_dir != "."
+    cd(call_dir)
+  end
+
+  return result # (out, err, ret)
+end
+
+# This function wraps a kbprog binary
+#     groupname::String   --- input file name
+#     dir::String         --- working directory
+function kbprog_call(groupname::String;
+                     dir::String = kbmag_data_dir)
+
+  res = kbmag_bin_wrapper("kbprog"; params = (groupname, ), force_dir = dir)
+
+  # assure that the call was successful
+  res.ret == 0 || error(res.err)
```

Comment:
Well, this is not that easy. First, `kbmag_bin_wrapper` is very generic. It is hard do say whether a non-zero exit status of any kbmag binary should entain an error. In case of single binaries such as `kbprog` or `wordreduce` we know the answer to this question and error handling is taken care of in proper `*_call` functions.

Second (or first, if we start from the top comment), `res.err` contains an error message, not the exit status. It is not helpful to know the status if the binary prints a specific message on `stderr`. Examples:
```
julia> KBmag.wordreduce_call("nonono", [""])
ERROR: Error: cannot open file nonono
```
or
```
julia> KBmag.wordreduce_call("237", ["nonono"])
ERROR: #Input error: invalid entry in word.
```

Reply:
very well;
could you add those two as tests? (there is `@test_throws` macro)

Reply:
Done.